Add embedded_hal::serial implementation for uarte (continuation)

hargoniX opened the pull request:
Since #281 has been abandoned by the author this is a continuation of his work in a seperate PR.

jonas-schievink commented on nrf-hal-common/src/uarte.rs:
```diff
@@ -406,18 +443,252 @@ pub enum Error {
+
+        // Prevent writing to buffer while DMA transfer is in progress.
+        if uarte.events_txstarted.read().bits() == 1 {
+            return Err(nb::Error::WouldBlock);
```
Where does `events_txstarted` get reset if we hit this path?

jonas-schievink reviewed (commented):
How was this code tested? This seems potentially tricky to get right, so I'm a bit hesitant to merge this without proper testing and review (I'd appreciate others to review this for correctness as well)

hargoniX commented:
I tested the UARTE with this little program on the micro:bit:
```rs
#![no_main]
#![no_std]

use panic_halt as _;

use core::fmt::Write;

use microbit::{
    hal::uarte,
    hal::uarte::{Baudrate, Parity},
    hal::prelude::*,
};

static mut TX_BUF: [u8; 1] = [0; 1];
static mut RX_BUF: [u8; 1] = [0; 1];

use cortex_m_rt::entry;


#[entry]
fn main() -> ! {
    let board = microbit::Board::take().unwrap();

    let (mut tx, mut rx) = {
        let serial = uarte::Uarte::new(
            board.UARTE0,
            board.uart.into(),
            Parity::EXCLUDED,
            Baudrate::BAUD115200,
        );
        serial.split(
            unsafe { &mut TX_BUF },
            unsafe { &mut RX_BUF }
        ).unwrap()
    };

    loop {
        write!(tx, "Hello World:\r\n").unwrap();
        let input = nb::block!(rx.read()).unwrap();
        write!(tx, "You wrote: {}\r\n", input as char).unwrap();
    }
}
```
As well as smaller and bigger variants of the TX_BUF size and it all seems to work. You were indeed right about the write not being functional though so I added some more logic that should make sure the flags are always properly reset and we only return WouldBlock if the TX transaction has both been started *and* is not finished yet.

jonas-schievink commented:
Yeah, I'm not entirely sure what the intended behavior from `embedded-hal` is there, but it seems fine to just write to a buffer for now. If this is wrong, the `embedded-hal` docs should be clarified to forbid this.

This PR still needs a rebase, but other than that I think it's ready to land. Thanks!

jonas-schievink commented:
bors r+

jonas-schievink commented:
bors retry

jonas-schievink commented:
bors r+
